pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -340,9 +340,18 @@ public void addSegment(String segmentName, TableConfig tableConfig, IndexLoading
       PartitionUpsertMetadataManager partitionUpsertMetadataManager =
           _tableUpsertMetadataManager != null ? _tableUpsertMetadataManager
               .getOrCreatePartitionManager(partitionGroupId) : null;
-      segmentDataManager =
-          new LLRealtimeSegmentDataManager(segmentZKMetadata, tableConfig, this, _indexDir.getAbsolutePath(),
-              indexLoadingConfig, schema, llcSegmentName, semaphore, _serverMetrics, partitionUpsertMetadataManager);
+      try {
+        segmentDataManager =
+            new LLRealtimeSegmentDataManager(segmentZKMetadata, tableConfig, this, _indexDir.getAbsolutePath(),
+                indexLoadingConfig, schema, llcSegmentName, semaphore, _serverMetrics, partitionUpsertMetadataManager);
+      } catch (Exception e) {
+        // In case of exception thrown here, segment goes to ERROR state. Then any attempt to reset the segment from
+        // ERROR -> OFFLINE -> CONSUMING via Helix Admin fails because the semaphore never gets released and the
+        // semaphore.acquire() is called in constructor of LLRealtimeSegmentDataManager. Hence releasing the semaphore
+        // here to unblock reset operation via Helix Admin.
+        semaphore.release();
+        throw e;
```

Comment:
It's little weird that the constructor is acquiring the semaphore, but looks much harder to change that here. 

I think ideally whatever work is being done there releases the semaphore. But considering we are going to throw anyway if something goes wrong in `LLRealtimeSegmentDataManager`, i guess there's no harm in doing this here.

Reply:
The `release()` should be handled within the constructor when it throws the exception. The problem of handling it here is that there is no guarantee `acquire()` is already called within the constructor

Reply:
Agreed. It is confusing to acquire in one place and release some place else. 

Perhaps a try/catch on all lines after the semahore (or any other resource) is acquired, so that if any exceptions are thrown, these can be released.

Reply:
IMO we should not rely on the fact that the semaphore is used as a mutex (1 entrant). Also, if by any chance 2 constructors run at the same time, one acquired the semaphore without exception, the other one will actually release the semaphore. This can be super rare, but just an example why we should ensure release happens after acquire

Reply:
@jackjlli introduced this semaphore a while ago, and IIRC we had a similar discussion. Code has changed significantly since then. Asymmetric acquire/release is not intuitive, I agree

Reply:
Semantically the two ways are the same. I didn't want to put a giant try block which spans more than 100 lines of code, but it's a valid point to release the semaphore next to the code where it's acquired. Will refactor the code shortly.

Reply:
Used to be that we never threw exception from this method. All the interaction with underlying stream was done in the thread that was spawned.  Well, now we like to try to do as much as possible in the state transition so that it can change to ERROR if we are not able to connect to the stream (for example). We need to change the semaphore etc. behavior accordingly.

You can move a whole bunch of lines into another method, and put a try/catch around a call to the method (if you think that gets to be more readable). Maybe throw specific exception also after catching an overall exception.

Reply:
+1 on wrapping the actual logic to another method to keep things simple.

Reply:
Carved out some logic in the constructor into smaller functions. Now try block is not that big. If it looks good to you, please approve & merge.